### PR TITLE
Structured metadata redux

### LINKED ISSUE #189: Update sweepers to support new structured metadata
### Are you sure this is not a new requirement or bug?

Yes

### 💡 Description

Update sweepers to support new structured metadata.

Related to
- https://github.com/NASA-PDS/harvest/issues/244
- https://github.com/NASA-PDS/registry-api/issues/632

### PR DESCRIPTION
**Pull request by alexdunnjpl**
## 🗒️ Summary
Updates sweepers to use new metadata structure, where flat /-separated keys are replaced by opensearch nested paths.

Disables repairkit sweeper as this is understood not to be a necessary part of sweepers anymore - this will be removed once this is confidently confirmed.

Renames the parent attribute from `ops:Provenance` to `ops:Registry_Sweepers`

Further update will be necessary for managed-opensearch migration, but that's downstream.

### 🤖 AI Assistance Disclosure
- [] No AI assistance used
- [X] AI used for light assistance (e.g., suggestions, refactoring, documentation help, minor edits)
- [ ] AI used for moderate content generation (AI generated some code or logic, but the developer authored or heavily revised the majority)
- [ ] AI generated substantial portions of this code

Estimated % of code influenced by AI: 5%

## ⚙️ Test Data and/or Report
Tests pass
Manually validated against docker registry with structured-metadata version of registry-loader

## ♻️ Related Issues
resolves #189 

## 🤓 Reviewer Checklist

*Reviewers: Please verify the following before approving this pull request.*

### Documentation and PR Content
- [ ] **Documentation:** README, Wiki, or inline documentation (Sphinx, Javadoc, Docstrings) have been updated to reflect these changes.
- [ ] **Issue Traceability:** The PR is linked to a valid GitHub Issue
- [ ] **PR Title:** The PR title is "user-friendly" clearly identifying what is being fixed or the new feature being added, that if you saw it in the Release Notes for a tool, you would be able to get the gist of what was done.

### Security & Quality
- [ ] **SonarCloud:** Confirmed no new High or Critical security findings.
- [ ] **Secrets Detection:** Verified that the Secrets Detection scan passed and no sensitive information (keys, tokens, PII) is exposed.
- [ ] **Code Quality:** Code follows organization style guidelines and best practices for the specific language (e.g., PEP 8, Google Java Style).

### Testing & Validation
- [ ] **Test Accuracy:** Verified that test data is accurate, representative of real-world PDS4 scenarios, and sufficient for the logic being tested.
- [ ] **Coverage:** Automated tests cover new logic and edge cases.
- [ ] **Local Verification:** (If applicable) Successfully built and ran the changes in a local or staging environment.

### Maintenance
- [ ] **Backward Compatibility:** Confirmed that these changes do not break existing downstream dependencies or API contracts (or that breaking changes are clearly documented).

### CODE REVIEW

File: src/pds/registrysweepers/repairkit/allarrays.py [hunk 1]
```diff
@@ -20,7 +20,7 @@
 
 def repair(document: Dict, fieldname: str) -> Dict:
     # don't touch the enumerated exclusions, or any registry-sweepers metadata property
-    if fieldname in EXCLUDED_PROPERTIES or fieldname.startswith("ops:Provenance"):
+    if fieldname in EXCLUDED_PROPERTIES or fieldname.startswith("ops:Registry_Sweepers"):
```
Comment (nutjob4life) [inline src/pds/registrysweepers/repairkit/allarrays.py:23]: This is great for flattened document keys 🎉 

But if documents are now structured like this:
```python
{
    "ops:Registry_Sweepers": {
        "ops:repairkit_software_version": 5
    }
}
```
then `allarrays.repair(document, fieldname)` may not even see nested fields unless upstream code flattens them before calling `repair`.

File: src/pds/registrysweepers/utils/db/indexing.py [hunk 2]
```diff
@@ -22,12 +22,17 @@ def _get_existing_mapping_type(client: OpenSearch, index_name: str, property_nam
...
         response.get(index_name, {})
         .get("mappings", {})
-        .get("properties", {})
     )
-    return existing_properties.get(property_name, {}).get("type")
+
+    # extract the relevant object, accounting for nesting
+    it = existing_mappings
+    for node in property_name.split('.'):
+        it = it.get('properties', {}).get(node, {})
```
Comment (nutjob4life) [inline src/pds/registrysweepers/utils/db/indexing.py:33]: If `it` ever stops being a dict, this could crash. Potentially safer:
```python
for node in property_name.split('.'):
    if not isinstance(it, dict):
        return None
    it = it.get("properties", {}).get(node, {})
```

File: src/pds/registrysweepers/utils/db/indexing.py [hunk 3]
```diff
@@ -70,16 +75,17 @@ def ensure_index_mapping(client: OpenSearch, index_name: str, property_name: str
...
         return
     except MissingMappingError:
         logger.info(
-            f"Mapping for '{property_name}' not yet present in index '{index_name}' - creating mapping with type {property_type}."
+            f"Mapping for '{property_key}' not yet present in index '{index_name}' - creating mapping with type {property_type}."
         )
         client.indices.put_mapping(
             index=index_name,
-            body={"properties": {property_name: {"type": property_type}}},
+            body={"properties": {property_key: {"type": property_type}}},
```
Comment (nutjob4life) [inline src/pds/registrysweepers/utils/db/indexing.py:88]: Double-checking: if `property_key` is now something like `ops:Registry_Sweepers.ops:reindexed_at`, OpenSearch will likely create a literal field with a **dot** in the name? Or will reject/interpret it depending on its version or configuration?

It **won't create** this:
```json
{
  "ops:Registry_Sweepers": {
    "properties": {
      "ops:reindexed_at": { "type": "date" }
    }
  }
}
```
But if _that_ ↑ is the goal, then we need a helper to convert dotted property paths into nested mapping bodies, sort of like
```python
def build_mapping_body(property_key: str, property_type: str) -> dict:
    parts = property_key.split(".")
    leaf = {"type": property_type}

    for part in reversed(parts[1:]):
        leaf = {"properties": {part: leaf}}

    return {"properties": {parts[0]: leaf}}
```
called via
```python
client.indices.put_mapping(
    index=index_name,
    body=build_mapping_body(property_key, property_type),
)
```

File: src/pds/registrysweepers/utils/db/multitenancy.py [hunk 4]
```diff
@@ -17,7 +17,10 @@ def resolve_multitenant_index_name(client: Union[OpenSearch, None], index_type:
     elif index_type not in supported_index_types:
         raise ValueError(f'index_type "{index_type}" not supported (expected one of {supported_index_types})')
     else:
-        return resolve_index_name_if_aliased(client, f"{node_id}-{index_type}")
+        # todo: determine whether or not this is an intended/permanent change - edunn 20260617
```
Comment (nutjob4life) [inline src/pds/registrysweepers/utils/db/multitenancy.py]: Can I admit a comment like this is sort of scary? 😬 

Other reviewers might even call it "merge blocking" 😮

File: src/pds/registrysweepers/utils/db/multitenancy.py [hunk 5]
```diff
@@ -17,7 +17,10 @@ def resolve_multitenant_index_name(client: Union[OpenSearch, None], index_type:
...
+        post_structured_index_type_name = index_type.replace('registry', 'registry-structured')
```
Comment (nutjob4life) [inline src/pds/registrysweepers/utils/db/multitenancy.py]: Probably safe although it does transform _any_ occurrence of `registry`, including substrings.

File: tests/pds/registrysweepers/repairkit/test_allarrays.py [hunk 6]
```diff
@@ -22,10 +22,10 @@ def test_exclusion_logic(self):
             "title": "Clementine LWIR brightness temperature flat field product: ff034ag.img",
             "product_class": "Product_Observational",
             "_package_id": "c0491371-49f8-4e34-9d1c-f94ef1217b57",
-            "ops:Provenance/ops:parent_collection_identifier": ["urn:nasa:pds:clementine_lwir_bt:data_flatfield::1.0"],
-            "ops:Provenance/ops:parent_bundle_identifier": ["urn:nasa:pds:clementine_lwir_bt::1.0"],
-            "ops:Provenance/ops:registry_sweepers_repairkit_version": 2,
-            "ops:Provenance/someStringTypedProp": "someValue",
+            "ops:Registry_Sweepers/ops:parent_collection_identifier": ["urn:nasa:pds:clementine_lwir_bt:data_flatfield::1.0"],
```
Comment (nutjob4life) [inline tests/pds/registrysweepers/repairkit/test_allarrays.py]: Should these slashes become dots?

**Review by nutjob4life — CHANGES_REQUESTED** [review]
- Schema migration: 🎉
- Some other issues: maybe, see the interspersed comments
- My Sweepers skills: low
- Defering to your expertise: absolutely

What do you think?

**Comment by jordanpadams** [comment]
@alexdunnjpl can you take a look at some of the comments provided by @nutjob4life above?

File: src/pds/registrysweepers/utils/db/multitenancy.py (hunk 5 above)
Comment (alexdunnjpl) [reply src/pds/registrysweepers/utils/db/multitenancy.py]: Substring replacement is intended (but ty!) - the latest version of harvest/registry-mgr replaces the canonical `registry*` index base names with `registry-structured*`

File: tests/pds/registrysweepers/repairkit/test_allarrays.py (hunk 6 above)
Comment (alexdunnjpl) [reply tests/pds/registrysweepers/repairkit/test_allarrays.py]: Will fix, tyvm

File: src/pds/registrysweepers/repairkit/allarrays.py (hunk 1 above)
Comment (alexdunnjpl) [reply src/pds/registrysweepers/repairkit/allarrays.py:23]: Repairkit is probably-dead code - @jordanpadams how should I handle repairkit?

 - delete the whole sweeper
 - delete all the obsolete rules and replace with a template/example, keep the sweeper structure
 - mark the files in self-doco as out-of-date, and requiring review should repairkit ever be needed again (I think we determined it shouldn't be now that harvest enforces versions?)
 - take the time to migrate/test repairkit properly instead of the current cursory attempt.

File: src/pds/registrysweepers/utils/db/indexing.py (hunk 2 above)
Comment (alexdunnjpl) [reply src/pds/registrysweepers/utils/db/indexing.py:33]: rgr, will fix, ty

File: src/pds/registrysweepers/utils/db/indexing.py (hunk 3 above)
Comment (alexdunnjpl) [reply src/pds/registrysweepers/utils/db/indexing.py:88]: @nutjob4life 

> It won't create this

Are you sure of this?  I'm 99.85% sure I tested this locally during development and found that OpenSearch handled dotted paths transparently even in index mapping calls.

About as important an issue as can exist, and simple for me to check, so I'll verify and report back here later.

File: src/pds/registrysweepers/utils/db/multitenancy.py (hunk 4 above)
Comment (alexdunnjpl) [reply src/pds/registrysweepers/utils/db/multitenancy.py]: Fair and fair 😅 

In this case, harvest/mgr changed the naming convention as part of the structured-schema changes, and this is coded to reflect that fact, but I couldn't see a good reason it should need to be named differently so wanted to leave a breadcrumb in case it got switched back later (once we back-migrate to managed OpenSearch and have to reindex everything fresh.

@jordanpadams should I leave this here, or accept that we're calling officially calling our indices `registry-structured*` going forward?

File: src/pds/registrysweepers/utils/db/indexing.py (hunk 3 above)
Comment (nutjob4life) [reply src/pds/registrysweepers/utils/db/indexing.py:88]: @alexdunnjpl I am 97.63% _unsure_ of that!

I'm more than willing to accept your expertise and testing of this 😁

**Comment by alexdunnjpl** [comment]
> @alexdunnjpl can you take a look at some of the comments provided by @nutjob4life above?

Sorry I missed these! Thanks Jordan

File: src/pds/registrysweepers/repairkit/allarrays.py (hunk 1 above)
Comment (jordanpadams) [reply src/pds/registrysweepers/repairkit/allarrays.py:23]: @alexdunnjpl if it totally doesn't work. let's wipe it. if it would require some upgrade but could still be useful in the future, let's document it as such, and make sure it cannot be executed.

Comment (jordanpadams) [reply src/pds/registrysweepers/repairkit/allarrays.py:23]: But def don't spend time on fixing it right now. We will address that when we need it.

Comment (alexdunnjpl) [reply src/pds/registrysweepers/repairkit/allarrays.py:23]: Rgr - I'll sprinkle `NotImplementedError` throws throughout with adjacent doco.

File: src/pds/registrysweepers/utils/db/multitenancy.py (hunk 4 above)
Comment (jordanpadams) [reply src/pds/registrysweepers/utils/db/multitenancy.py]: @alexdunnjpl Let's just assume that to be true. I am not intimately involved in this implementation so I really don't know, but we need to keep making progress. Pick something and let's just make sure we are paying attention when this reindexing/migration occurs.

**Comment by jordanpadams** [comment]
@alexdunnjpl one more thing: how can we test this?

**Comment by alexdunnjpl** [comment]
@jordanpadams for regression, initial validation, or how can a reviewer test this PR?

Regression should be covered (to the extent we've historically been covered) by I&T tests, once I&T is using the structured versions of api/harvest/etc and sweepers.

I validated this manually, by side-loading a locally-built registry-loader image as of `11c3ba46` into registry, checking that the loaded data was structured as expected, then running registry sweepers as of `bdabc29` against the dockerized registry and confirming correct execution.

Someone wanting to verify would need to do the same.

If the API has been updated already, a better option (and something that I can do myself) would be to build bleeding-edge images for registry-loader, registry-sweepers, *and* registry-api, and run the docker I&T profile with those images side-loaded.

**Comment by jordanpadams** [comment]
@alexdunnjpl any updates on getting the fixes in recommended by Sean above?

File: src/pds/registrysweepers/utils/db/indexing.py (hunk 2 above)
Comment (alexdunnjpl) [reply src/pds/registrysweepers/utils/db/indexing.py:33]: Belay that - if if ever stops being a dict, it *should* crash, because this content is coming straight from opensearch's representation of the mappings and that would indicate a massive problem with opensearch, or a bad assumption baked into our code.

**Comment by nutjob4life** [comment]
@alexdunnjpl this is fantastic. Thanks for nursing this along. We are almost there!

I think one thing _other_ devs might not expect is that `tox` reports failures:
```console
$ git checkout structured-metadata-redux
Switched to branch 'structured-metadata-redux'
Your branch is up to date with 'origin/structured-metadata-redux'.
$ python3 -m venv .venv
$ . .venv/bin/activate
(.venv) $ pip install --editable --quiet '.[dev]'
(.venv) $ tox
…
  py312: FAIL code 1 (16.05=setup[11.33]+cmd[4.72] seconds)
  py313: FAIL code 1 (15.99=setup[11.18]+cmd[4.81] seconds)
  docs: OK (10.92=setup[9.69]+cmd[1.24] seconds)
  lint: OK (17.97=setup[1.37]+cmd[16.60] seconds)
  evaluation failed :( (61.17 seconds)
```
Of course what fails during the "test" phase is:
```
FAILED tests/pds/registrysweepers/repairkit/test_allarrays.py::AllArrays::test_invalid_field - NotImplementedError: The repairkit sweeper is deprecated and must be updated/tested for structured-property compati...
FAILED tests/pds/registrysweepers/repairkit/test_allarrays.py::AllArrays::test_exclusion_logic - NotImplementedError: The repairkit sweeper is deprecated and must be updated/tested for structured-property compati...
FAILED tests/pds/registrysweepers/repairkit/test_allarrays.py::AllArrays::test_valid_field - NotImplementedError: The repairkit sweeper is deprecated and must be updated/tested for structured-property compati...
```
which _we_ expect, but other devs (and automated processes) might not comprehend.

Maybe we can mark those tests as disabled for now? Like `@pytest.mark.skip` or `@pytest.mark.skipif` or something?

**Comment by alexdunnjpl** [comment]
@tloubrieu-jpl please review https://github.com/NASA-PDS/registry-sweepers/pull/242/changes/f55c9630cc5119af1ca8934c3cd372e2bd3311e1 and confirm I'm not making any bad assumptions about how the legacy sync is supposed to work?

It seems straightforward, but I don't understand how it was working in multitenant AOSS before my changes without applying the appropriate node prefix.

@jordanpadams FYI

**Comment by alexdunnjpl** [comment]
@nutjob4life my bad my bad - I've removed the injudicious throw.  Since we may want to use this functionality in future, best that the regression tests for the internals remain.

**Review by nutjob4life — APPROVED** [review]
Thanks for putting up with all my pedantry!

Approval: ✅✅✅
Tox details:

Tox output: success
